website/registrations/utils.py

```diff
@@ -0,0 +1,349 @@
+                   'license': {'format': 'object', 'required': True},
+                   'subjects': {'format': 'list', 'required': True},
+                   'tags': {'format': 'list'},
+                   'project guid': {'format': 'string'},
+                   'external id': {'format': 'string'}}
+CONTRIBUTOR_METADATA_FIELDS = ['admin', 'read-write', 'read-only', 'bibliographic contributors']
+
+class BulkRegistrationUpload():
+    @property
+    def is_valid(self):
+        return not len(self.errors)
+
+    def __init__(self, bulk_upload_csv, provider_id):
+        self.raw_csv = bulk_upload_csv.read()
+        # self.raw_csv = bulk_upload_csv.read().decode('utf-8')
+        self.reader = csv.DictReader(io.StringIO(self.raw_csv))
+        self.headers = [header.lower() for header in self.reader.fieldnames]
+        schema_id_row = next(self.reader)
+        self.schema_id = schema_id_row[self.reader.fieldnames[0]]
+        self.provider_id = provider_id
+        if self.schema_id:
+            try:
+                self.registration_provider = RegistrationProvider.load(self.provider_id)
+                self.registration_schema = self.registration_provider.schemas.get(_id=self.schema_id)
+            except RegistrationSchema.DoesNotExist:
+                raise NotFound(detail='Schema with id "{}" was not found'.format(self.schema_id))
+            except RegistrationProvider.DoesNotExist:
+                raise NotFound(detail='Registration provider with id "{}" was not found').format(self.provider_id)
+        self.schema_questions = BulkRegistrationUpload.get_questions_validations(self.registration_schema)
+        self.validations = {**self.schema_questions, **METADATA_FIELDS}
+        self.errors = []
+        self.rows = [Row(row,
+                         self.validations,
+                         functools.partial(self.log_error, row_index=self.reader.line_num))
+                     for row in self.reader]
+
+    @classmethod
+    def get_questions_validations(cls, registration_schema):
+        schema_questions = {}
+        get_question_validations = lambda question: {'required': question.get('required', False),
+                                   'format': question.get('format', ''),
+                                   'type': question.get('type', 'string'),
+                                   'options': question.get('options', [])}
+        for page in registration_schema.schema['pages']:
+            for question in page['questions']:
+                schema_questions[question['qid']] = get_question_validations(question)
+                if 'properties' in question:
+                    for nested_question in question['properties']:
+                        qid = '{}{}'.format(question['qid'], nested_question['id'])
+                        schema_questions[qid] = get_question_validations(nested_question)
+        return schema_questions
+
+    def log_error(self, **kwargs):
```

Comment:
According to the error doc provided by Mark, the information included here is probably not enough to generate the same error msg as those he requested. We cannot tell what type of fields the errors are attached to. Also, the format he suggested is an aggregate format, with one error msg for all the cells with the same error type.

Reply:
addressed in flow